### ISSUE TITLE
InMemoryFactory on a cluster environment

### ISSUE THREAD

**Issue by imperugo** [thinktecture/Thinktecture.IdentityServer.v3#786]
Hi, 
I'd like to know if I can use InMemoryFactory in a cluster environment (2 or more servers).

Basically for me InMemory doesn't fit so good with distributed systems but maybe the factory doesn't need to store information.

So, should I create a Factory using Redis/Sql on InMemory is good?

**Comment by brockallen** [comment]
The docs discuss this: https://thinktecture.github.io/Thinktecture.IdentityServer.v3.Documentation/docs/configuration/serviceFactory.html

**Comment by imperugo** [comment]
Great.

Thanks

**Comment by pnally** [comment]
These docs are not available anymore. My main question is this:
I don't want to use Entity Framework for the persistence layer of IDSVR3, so is there documentation or example code that will allow me to create a custom persistence layer for the "Operational Data" discussed in this documentation?
[https://identityserver.github.io/Documentation/docsv2/advanced/deployment.html](https://identityserver.github.io/Documentation/docsv2/advanced/deployment.html)

I feel like this question has the same issue as mine, being that we are using inmemory factory, so I'm trying to break each component out one by one to minimize the points of in memory persistence for each. scopes and clients should be fine in memory, because the definition of those don't change, but maybe I'm misunderstanding a fundamental definition of clients and scopes.

I keep getting referred to this link for my IndentityServer3 implementation regarding getting identity server to work in a cluster setup and, frankly, it's not that helpful on helping me with the specific options I have for implementing a custom persistence layer for the operational data: [https://identityserver.github.io/Documentation/docsv2/advanced/deployment.html](https://identityserver.github.io/Documentation/docsv2/advanced/deployment.html) 

I'm hoping to get direction on what methods and/or configurations need to be implemented in order to store operational data in a SQL server database (the same database we are using to store our user data).

The only solution we have right now is to set the load balancer to force same server based on IP address, or take the identity server out of the cluster.  We are getting a "An error has occurred" when requesting the login page coming from a javascript frontend using the oidc-client.js framework.  We are using inmemory scope and client, but a custom persistence layer for users.  The problem seems to be when requesting the authorization tokens.

Anyway, I really appreciate any guidance anyone has to point me in the right direction. Thanks in advance.

**Comment by brockallen** [comment]
To build your own, I'd suggest taking a look at how it's done in the EF implementations. From that guidance, then build your own.

**Comment by pnally** [comment]
Thank you for confirming my direction (it's what I've been doing). 

**What I've already done:** I've used the CustomUserService example to produce the identity server I'm using, and now trying to implement the OperationalDbContext found in the EntityFramework example. Trying to understand/change the EF version for my custom db solution. But now when I'm debugging the solution _(after creating the custom Factory and the respective stores / classes / entities / interfaces / IdentityServerServiceFactoryExtensions / etc. related to the OperationalDbContext)_ the context of the BaseTokenStore is always null for the Consents, Schema, and Tokens. I figured the clients and scopes are always statically defined in the clients and stores config classes and don't need to be stored in the persistence layer, so I'm still adding in the in-memory version of those when creating the custom factory. 

**My Question:** So why would the BaseTokenStore be empty each time the login is requested? Where in the EF examples does the overriding of the creation and more importantly the storage of the the tokens that ultimately need to be shared in the persistence later?  I thought it was the 'StoreAsync' methods defined in the respective stores (AuthorizationCodeStore, ConsetStore, RefreshTokenStore, TokenHandleStore) that I should change to make custom calls to my own persistence later, is that not the correct location for that?
